File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const n = nums.length;

    // Initialize the result array with all 1s
    const result = Array(n).fill(1);

    let prefix = 1;
    
    // Calculate the prefix products and store them in the result array
    for (let i = 0; i < n; i++) {
        result[i] *= prefix;
        prefix *= nums[i];
    }

    let suffix = 1;

    // Update the result array with the suffix products
    for (let i = n - 1; i >= 0; i--) {
        result[i] *= suffix;
        suffix *= nums[i];
    }

    return result;
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [1] * n

        prefix, suffix = 1, 1

        for i in range(n):
            result[i] *= prefix
            result[n - 1 - i] *= suffix
            prefix *= nums[i]
            suffix *= nums[n - 1 - i]

        return result
